src/index/IndexMetaData.h

```diff
@@ -200,19 +81,30 @@ class IndexMetaData {
   size_t getNofDistinctC1() const;
 
  private:
-  off_t _offsetAfter;
+  off_t _offsetAfter = 0;
   size_t _nofTriples;
+
   string _name;
+  string _filename;
 
-  ad_utility::HashMap<Id, FullRelationMetaData> _data;
+  MapType _data;
   ad_utility::HashMap<Id, BlockBasedRelationMetaData> _blockData;
 
+  
+  // this way all instantations will be friends with each other, 
+  // but this should not be an issue.
+  template<class U>
   friend ad_utility::File& operator<<(ad_utility::File& f,
-                                      const IndexMetaData& rmd);
+                                      const IndexMetaDataTemplated<U>& rmd);
 
   size_t getNofBlocksForRelation(const Id relId) const;
 
   size_t getTotalBytesForRelation(const FullRelationMetaData& frmd) const;
 };
 
-ad_utility::File& operator<<(ad_utility::File& f, const IndexMetaData& imd);
\ No newline at end of file
+// aliases for easier use in Index class
+using IndexMetaDataExtVec = IndexMetaDataTemplated<MetaDataWrapperExtVec>;
+using IndexMetaData = IndexMetaDataTemplated<MetaDataWrapperHashMap>;
+using IndexMetaDataMmap = IndexMetaDataTemplated<MetaDataWrapperMmap<MmapBasedArray<FullRelationMetaData>>>;
```

Comment:
Shouldn't `MetaDataWrapperMmap` for `MmapBasedArray` be just a wrapper for random access containers that also works with `std::array` and `std::vector`? If not which features are specific to `MmapBasedArray`?

Reply:
I fixed it and renamed in to a MetaDataWrapperDense
The only thing which has to be taken care of is the way we read and write from disk.
For persistent data types (like MmapVector) it suffices to read and write the block data, otherwise you also have to write the FullRelationalMetaData. But since the latter always works and is the default, we should be fine.